File: src/fbclient/src/input.zig
/// Input abstraction: maps platform keys to app-level Key enum.
/// Backend: SDL2 events (dev) or evdev (device).
const std = @import("std");
const build_options = @import("build_options");

pub const Key = enum {
    up,
    down,
    left,
    right,
    enter, // SELECT / DPAD_CENTER
    back, // BACK / ESC
    ptt, // F1 — always routed to wata
    dot1, // F3 — prev applet
    dot2, // F6 — next applet
    f2, // F2 — spare
    unknown,
};

pub const KeyState = enum { pressed, released, repeat };

pub const InputEvent = struct {
    key: Key,
    state: KeyState,
};

pub const Backend = if (build_options.use_sdl) SdlInput else EvdevInput;

// ---------------------------------------------------------------------------
// SDL2 input
// ---------------------------------------------------------------------------
const SdlInput = struct {
    const sdl = @import("sdl.zig").c;

    pub fn init() !SdlInput {
        return .{};
    }
    pub fn deinit(_: *SdlInput) void {}

    /// Poll all pending events. Returns a slice of InputEvents (up to max_events).
    /// Also returns true for `quit` if the window is closed.
    pub fn poll(self: *SdlInput, buf: []InputEvent) struct { events: []InputEvent, quit: bool } {
        _ = self;
        var count: usize = 0;
        var quit = false;
        var ev: sdl.SDL_Event = undefined;

        while (sdl.SDL_PollEvent(&ev) != 0) {
            if (ev.type == sdl.SDL_QUIT) {
                quit = true;
                continue;
            }
            if (ev.type == sdl.SDL_KEYDOWN or ev.type == sdl.SDL_KEYUP) {
                const state: KeyState = if (ev.type == sdl.SDL_KEYDOWN)
                    (if (ev.key.repeat != 0) .repeat else .pressed)
                else
                    .released;

                const key: Key = switch (ev.key.keysym.sym) {
                    sdl.SDLK_UP => .up,
                    sdl.SDLK_DOWN => .down,
                    sdl.SDLK_LEFT => .left,
                    sdl.SDLK_RIGHT => .right,
                    sdl.SDLK_RETURN, sdl.SDLK_KP_ENTER => .enter,
                    sdl.SDLK_ESCAPE, sdl.SDLK_BACKSPACE => .back,
                    sdl.SDLK_SPACE => .ptt,
                    sdl.SDLK_F1 => .ptt,
                    sdl.SDLK_F2 => .f2,
                    sdl.SDLK_F3 => .dot1,
                    sdl.SDLK_F6 => .dot2,
                    // Tab and shift-tab as dot1/dot2 for convenience
                    sdl.SDLK_TAB => if ((sdl.SDL_GetModState() & sdl.KMOD_SHIFT) != 0) .dot1 else .dot2,
                    else => .unknown,
                };

                if (key != .unknown and count < buf.len) {
                    buf[count] = .{ .key = key, .state = state };
                    count += 1;
                }
            }
        }

        return .{ .events = buf[0..count], .quit = quit };
    }
};

// ---------------------------------------------------------------------------
// evdev input (Linux device)
// ---------------------------------------------------------------------------
const EvdevInput = struct {
    // Linux input event struct (matches kernel's struct input_event)
    const LinuxInputEvent = extern struct {
        tv_sec: isize, // adapts to 32-bit (ARM) or 64-bit
        tv_usec: isize,
        type: u16,
        code: u16,
        value: i32,
    };

    // Event types
    const EV_KEY: u16 = 0x01;

    // Key codes (from linux/input-event-codes.h, matching HARDWARE.md)
    const KEY_ESC: u16 = 1; // Matrix keypad: Back
    const KEY_ENTER: u16 = 28; // Matrix keypad: Center
    const KEY_F1: u16 = 59; // GPIO: Main PTT
    const KEY_F2: u16 = 60; // GPIO: Headset PTT
    const KEY_F3: u16 = 61; // GPIO: Side button 3 (prev applet)
    const KEY_F4: u16 = 62; // PMIC: RESIN button
    const KEY_F10: u16 = 68; // GPIO: Side button 4 (next applet)
    const KEY_UP: u16 = 103; // Matrix keypad
    const KEY_LEFT: u16 = 105;
    const KEY_RIGHT: u16 = 106;
    const KEY_DOWN: u16 = 108;

    const linux = std.os.linux;
    const MAX_FDS = 3;

    fds: [MAX_FDS]std.posix.fd_t,
    fd_count: u8,

    pub fn init() !EvdevInput {
        var self = EvdevInput{
            .fds = .{ -1, -1, -1 },
            .fd_count = 0,
        };

        // Open all input devices — skip any that fail to open
        const paths = [_][*:0]const u8{
            "/dev/input/event0", // PMIC PON (power, RESIN)
            "/dev/input/event1", // Matrix keypad (d-pad, enter, esc)
            "/dev/input/event2", // GPIO keys (PTT, F2, F3, F10)
        };

        for (paths) |path| {
            const fd = std.posix.openatZ(std.posix.AT.FDCWD, path, .{ .ACCMODE = .RDONLY, .NONBLOCK = true }, 0) catch continue;
            self.fds[self.fd_count] = fd;
            self.fd_count += 1;
        }

        if (self.fd_count == 0) return error.NoInputDevices;
        return self;
    }

    pub fn deinit(self: *EvdevInput) void {
        for (self.fds[0..self.fd_count]) |fd| {
            _ = linux.close(fd);
        }
    }

    pub fn poll(self: *EvdevInput, buf: []InputEvent) struct { events: []InputEvent, quit: bool } {
        var count: usize = 0;

        for (self.fds[0..self.fd_count]) |fd| {
            // Drain all available events from this device
            while (count < buf.len) {
                var ev: LinuxInputEvent = undefined;
                const n = std.posix.read(fd, std.mem.asBytes(&ev)) catch break; // WouldBlock → break
                if (n != @sizeOf(LinuxInputEvent)) break;

                if (ev.type != EV_KEY) continue;

                const state: KeyState = switch (ev.value) {
                    0 => .released,
                    1 => .pressed,
                    2 => .repeat,
                    else => continue,
                };

                const key = mapKey(ev.code);
                if (key == .unknown) continue;

                buf[count] = .{ .key = key, .state = state };
                count += 1;
            }
        }

        return .{ .events = buf[0..count], .quit = false };
    }

    fn mapKey(code: u16) Key {
        return switch (code) {
            KEY_UP => .up,
            KEY_DOWN => .down,
            KEY_LEFT => .left,
            KEY_RIGHT => .right,
            KEY_ENTER => .enter,
            KEY_ESC => .back,
            KEY_F1 => .ptt, // Main PTT (side)
            KEY_F2 => .f2, // Headset PTT (side)
            KEY_F3 => .dot1, // Side button 3 → prev applet
            KEY_F10 => .dot2, // Side button 4 → next applet (HW code is KEY_F10)
            else => .unknown,
        };
    }
};

File: wata-fb/src/main/scala/input.scala
import language.experimental.saferExceptions

/** evdev input. Reads `struct input_event`s from `/dev/input/event{0,1,2}`
 *  (nonblocking) and maps the kernel key codes (linux/input-event-codes.h) to
 *  the app `Key` enum. The `Key`/`KeyState` "enums" are SEALED nullary
 *  case-class families, not Scala 3 `enum`s — matching the same style used
 *  for `ConnectionState` in the portable domain model. Device-layer app code
 *  over the `go.syscall` facade. */

sealed trait Key derives CanEqual
case class KUp() extends Key
case class KDown() extends Key
case class KLeft() extends Key
case class KRight() extends Key
case class KEnter() extends Key      // SELECT / DPAD_CENTER
case class KBack() extends Key       // BACK / ESC
case class KPtt() extends Key        // F1 — main PTT
case class KDot1() extends Key       // F3 — prev applet
case class KDot2() extends Key       // F10 — next applet
case class KF2() extends Key         // F2 — sim/script only; no such key on the case
case class KUnknown() extends Key

sealed trait KeyState derives CanEqual
case class Pressed() extends KeyState
case class Released() extends KeyState
case class Repeat() extends KeyState

case class KeyEvent(key: Key, state: KeyState)

object Evdev:
  // struct input_event field/type constants
  val EV_KEY: scala.Int = 0x01
  val EVENT_SIZE: scala.Int = 16   // 32-bit ARM: tv_sec(4) tv_usec(4) type(2) code(2) value(4)

  // key codes (linux/input-event-codes.h)
  val KEY_ESC: scala.Int = 1       // Matrix keypad: Back
  val KEY_ENTER: scala.Int = 28    // Matrix keypad: Center
  val KEY_F1: scala.Int = 59       // GPIO: Main PTT
  val KEY_F2: scala.Int = 60       // GPIO: headset PTT line (no key on the case)
  val KEY_F3: scala.Int = 61       // GPIO: Side button 3 (prev applet)
  val KEY_F10: scala.Int = 68      // GPIO: Side button 4 (next applet)
  val KEY_UP: scala.Int = 103
  val KEY_LEFT: scala.Int = 105
  val KEY_RIGHT: scala.Int = 106
  val KEY_DOWN: scala.Int = 108

  def mapKey(code: scala.Int): Key =
    if code == KEY_UP then KUp()
    else if code == KEY_DOWN then KDown()
    else if code == KEY_LEFT then KLeft()
    else if code == KEY_RIGHT then KRight()
    else if code == KEY_ENTER then KEnter()
    else if code == KEY_ESC then KBack()
    else if code == KEY_F1 then KPtt()
    else if code == KEY_F2 then KF2()
    else if code == KEY_F3 then KDot1()
    else if code == KEY_F10 then KDot2()
    else KUnknown()

  def mapState(value: scala.Int): KeyState =
    if value == 1 then Pressed()
    else if value == 2 then Repeat()
    else Released()

  def known(k: Key): Boolean = k match
    case KUnknown() => false
    case _          => true

  def keyName(k: Key): String = k match
    case KUp()      => "up"
    case KDown()    => "down"
    case KLeft()    => "left"
    case KRight()   => "right"
    case KEnter()   => "enter"
    case KBack()    => "back"
    case KPtt()     => "ptt"
    case KDot1()    => "dot1"
    case KDot2()    => "dot2"
    case KF2()      => "f2"
    case KUnknown() => "unknown"

  def stateName(s: KeyState): String = s match
    case Pressed()  => "pressed"
    case Released() => "released"
    case Repeat()   => "repeat"

  def u8(buf: go.Bytes, i: scala.Int): scala.Int = buf(i).toInt & 0xff

  /** open /dev/input/event0..2 nonblocking; skip any that fail to open (a dev
   *  host may have none). Returns the open fds. */
  def open(): List[scala.Int] =
    var acc: List[scala.Int] = Nil
    var i = 0
    while i < 3 do
      val path = "/dev/input/event" + i
      try
        val fd = go.syscall.open(path, go.syscall.O_RDONLY | go.syscall.O_NONBLOCK, 0)
        acc = fd :: acc
      catch case e: sgo.GoError => ()
      i += 1
    ListOps.reverse(acc)

  def count(fds: List[scala.Int]): scala.Int =
    var n = 0
    var cur = fds
    var going = true
    while going do
      cur match
        case f :: t => n = n + 1; cur = t
        case Nil  => going = false
    n

  def closeAll(fds: List[scala.Int]): Unit =
    var cur = fds
    var going = true
    while going do
      cur match
        case fd :: t => go.syscall.close(fd); cur = t
        case Nil   => going = false

  /** poll all fds, draining every pending input_event; map to KeyEvents. */
  def poll(fds: List[scala.Int]): List[KeyEvent] =
    var acc: List[KeyEvent] = Nil
    var cur = fds
    var going = true
    while going do
      cur match
        case fd :: t =>
          acc = drainFd(fd, acc)
          cur = t
        case Nil => going = false
    ListOps.reverse(acc)

  /** drain one fd until EAGAIN (read throws) or a short read; prepend mapped
   *  key events onto `accIn`. */
  def drainFd(fd: scala.Int, accIn: List[KeyEvent]): List[KeyEvent] =
    var acc = accIn
    val buf = go.makeSlice[Byte](EVENT_SIZE)
    var reading = true
    while reading do
      try
        val n = go.syscall.read(fd, buf)
        if n != EVENT_SIZE then reading = false
        else
          val etype = u8(buf, 8) | (u8(buf, 9) << 8)
          if etype == EV_KEY then
            val code = u8(buf, 10) | (u8(buf, 11) << 8)
            val value = u8(buf, 12)
            val k = mapKey(code)
            if known(k) then acc = KeyEvent(k, mapState(value)) :: acc
      catch case e: sgo.GoError => reading = false
    acc
